### ISSUE TITLE
Split-screen external search and annotation view

### ISSUE THREAD

**Issue by rgabbard** [inception-project/inception#825]
**Is your feature request related to a problem? Please describe.**

The current workflow for external search splits searching and annotation into two phases: first you search for documents and import them, then you go open them for annotation.  This is inconvenient for workflows which depend on a faster back-and-forth between search and annotation.

**Describe the solution you'd like**

I would like to have a two-paned view, where a narrow left-hand pane had the search term box and the right-hand pane had the usual annotation interface open to the selected document. Ideally, the search query would be communicated to the annotation view in order to automatically scroll to the first matching portion of the document.

![img_20190104_130536 1](https://user-images.githubusercontent.com/7139661/50703430-1e9c4100-1022-11e9-8b49-7fd026d33751.jpg)

Details:
* presumably a document would be automatically imported into the document repository when its search result was clicked

**Comment by rgabbard** [comment]
(if we could implement this, it would remove the need for #792 and #793 as far as our particular project was concerned)

**Comment by shinyichen** [comment]
Would this page only handle external searches, leaving the internal search and annotation page the way it is? If so I think it sounds feasible.

I like this idea. Just throwing out some thoughts that came to mind:
1. After a search and you start to annotate, which I am not sure how long that usually takes, you probably want to maintain this list so you can come back to. Say if you navigate away from search, want to perform another search or want to come back to work on it later. Although you could just perform the same search again (from my observation it seems the results and orderings are always the same with the search), this workflow seems odd. I'd like to come back to the same set of documents, but I don't want to assume a search always returns the same results in same order. Or we don't really care about that?
2. Would it be confusing to have two places to annotate (from external search page and from annotation page).

**Comment by rgabbard** [comment]
@shinyichen : As far as ISI is concerned, we only care about external search.  Someone else might want to extend it to internal search in the future.

re: 1 - I think the idea is that the search result list on the left pane is persistent - so you click one document, annotate it in the right pane, click to the next document in the left pane, etc.

re: 2 - I imagine rather than replacing the existing interfaces (which work well for many applications), we would want to add some sort of new option for the hybrid interface

**Comment by rgabbard** [comment]
(I should note this is the sort of interface we used when we implemented the original curated training annotation tool)

**Comment by reckart** [comment]
> re: 2 - I imagine rather than replacing the existing interfaces (which work well for many applications), we would want to add some sort of new option for the hybrid interface

With "existing" you mean the full-screen search pane?

I imagine that implementing a custom sidebar which offers a simplified access to external search should be pretty straight-forward to do. Sidebars can already be added to the annotation page via an extension mechanism. Cf. e.g. the `SearchAnnotationSidebar` in the `inception-ui-search` module.

Persistence when switching across documents should work out-of-the-box. However, in order to make the results persistent when switching back and forth between different sidebars needs extra work. In the `ActiveLearningSidebar` we solved this by attaching the state of the sidebar directly to the surrounding `AnnotationPage` instance instead of maintaining it in the sidebar itself. Cf. the following code copy/pasted from `ActiveLearningSidebar`.

```
        // Instead of maintaining the AL state in the sidebar, we maintain it in the page because
        // that way we persists even if we switch to another sidebar tab
        alStateModel = new CompoundPropertyModel<>(LambdaModelAdapter.of(
            () -> aAnnotationPage.getMetaData(CURRENT_AL_USER_STATE),
            alState -> aAnnotationPage.setMetaData(CURRENT_AL_USER_STATE, alState)));
```

**Comment by rgabbard** [comment]
@reckart : thanks! @shinyichen , can you follow up @reckart 's pointers and write up a plan for this?

**Comment by rgabbard** [comment]
@reckart : yes, `existing` meant the current full-screen search pane.

**Comment by reckart** [comment]
Btw. the code in the `SearchAnnotationSidebar` illustrates how to jump to specific locations in documents and how to add match highlights. A quick start could be to copy the existing `SearchAnnotationSidebar` sidebar and then adjust it to query externally instead of internally. 

An (probably future) enhancement would be to query the internal and external services in parallel and to show results from both in the same sidebar.

**Comment by shinyichen** [comment]
Combining the above info regarding search persistence and sidebar, this is what I gather it would look like.
![img_20190104_133527](https://user-images.githubusercontent.com/636361/50712346-03cbcb80-1026-11e9-9afb-1b731ae2b1b2.jpg)

**Comment by rgabbard** [comment]
@shinyichen : For the moment, I don't think we need to worry about keeping past searches. When @reckart is talking above about keeping results persistent, I think he means that if you switch from the search sidebar to some other sidebar and then go back to the search sidebar, your search results should still be there.

**Comment by shinyichen** [comment]
Ok, but there are no other sidebars besides search in the initial drawing. What about persistence when we leave the page, do we want that?

**Comment by rgabbard** [comment]
I assume the annotation interface itself allows you to open other sidebars, potentially.

For the moment, I don't think we need to worry about persistence with respect to searches in any way at all except the sidebar-switching persistence noted above.  Some sort of search history could be tackled in the future in another issue.

**Comment by reckart** [comment]
> @reckart is talking above about keeping results persistent, I think he means that if you switch from the search sidebar to some other sidebar and then go back to the search sidebar, your search results should still be there.

If you switch between sidebars, then they normally discard their state because Wicket discards the UI components of invisible sidebars and re-creates them when they become visible again. On the `ActiveLearningSidebar`, this default behavior had the effect that the AL session was canceled whenever the user switch to another sidebar and then back to AL. We circumvented that by storing the AL state not in the AL sidebar but instead attaching it to the annotation page. If the user leaves the page and then goes back to the annotation page, the state is lost. It would be possible to attach states to the user's session which would make it persistent across pages, but I don't think it is a good idea to carry such a global state around.

**Comment by reckart** [comment]
The way the sidebar mechanism is currently designed, it is not possible to add multiple instances of the same sidebar (i.e. multiple `ExternalSearchSidebar`s). If you wanted to manage multiple searches and easily switch between them, then the approach @shinyichen illustrated of having a list of "open searches" within the sidebar makes sense (at least that is how I understood "search 1", "search 2"). But I believe it would get quite tight in terms of screen space ;)

**Comment by rgabbard** [comment]
@reckart, @shinyichen : I don't think handling multiple searches simultaneously is needed for ISI's application, so we can leave that for someone else to look into in the future.

**Comment by shinyichen** [comment]
After reading over the comments again, I understand what @reckart meant. You are talking about adding a new sidebar to the existing Annotation page, something like this. Now that makes more sense about the persistence.

<img width="1188" alt="screen shot 2019-01-04 at 2 20 54 pm" src="https://user-images.githubusercontent.com/636361/50714614-59a47180-102e-11e9-9b2c-8c2e533d8349.png">

**Comment by reckart** [comment]
@shinyichen exactly. 

I assume there should not be an "import" button in this case and instead the document should be auto-imported when the user clicks on a result (unless the document is already present in the project).

@rgabbard I wonder: if there are multiple hits in a document, do you expect there be a results row for each hit - e.g. one result row per matching sentence?

**Comment by shinyichen** [comment]
Here a write up from the information gathered.

## The split pane should have ##
- **External search** (left top) - Select a repo and search
- **Search results** (left) - Click on a row (result) will import the document and show text in annotation page
- **Annotation pane** (right)

## Implementation ##
We can add a new side bar to the existing annotation page.

![](https://d2mxuefqeaa7sj.cloudfront.net/s_A9076E752FDFE7AF0BF64082D18370B2E3151D0A5F4FCC2CECD8A6B6C71AA73B_1546642679237_Screen+Shot+2019-01-04+at+2.53.39+PM.png)


### Adding sidebar to annotation page ###
Sidebars can already be added to the annotation page via an extension mechanism. See example `SearchAnnotationSidebar` in the `inception-ui-search` module. A quick start could be to copy the existing `SearchAnnotationSidebar` sidebar and then adjust it to query externally.

### Add external search to sidebar ###
See current external search UI and implementation

### Search bar persistence ###
To save the state of the search while switching between sidebars, see `ActiveLearningSidebar`. Attaching the state of the sidebar directly to the surrounding `AnnotationPage` instance instead of maintaining it in the sidebar itself. Cf. the following code copy/pasted from `ActiveLearningSidebar`.
```
// Instead of maintaining the AL state in the sidebar, we maintain it in the page because
// that way we persists even if we switch to another sidebar tab
alStateModel = new CompoundPropertyModel<>(LambdaModelAdapter.of(
    () -> aAnnotationPage.getMetaData(CURRENT_AL_USER_STATE),
    alState -> aAnnotationPage.setMetaData(CURRENT_AL_USER_STATE, alState)));
```

### Other notes ###
The code in the `SearchAnnotationSidebar` illustrates how to jump to specific locations in documents and how to add match highlights.

**Comment by rgabbard** [comment]
@reckart : Different users would probably have different expectations with respect to multiple hits from the same document. For our use case, we would prefer each document appear only once, even if it has multiple matches.

**Comment by rgabbard** [comment]
@shinyichen : this looks like a good plan. Go ahead and start the implementation.

**Comment by shinyichen** [comment]
@reckart It doesn't seem like the external search (ElasticSearch) results have keyword markers. Is it yet to be implemented?

**Comment by reckart** [comment]
@shinyichen there are markers, but they are currently not explicitly modeled in the search results. Here you can see the code that handles the markers and displays them on the external search page:

https://github.com/inception-project/inception/blob/7c50defbfe6f941692b72a535c2021692fa1caf5/inception-ui-external-search/src/main/java/de/tudarmstadt/ukp/inception/app/ui/externalsearch/SearchPage.java#L309-L315

**Comment by shinyichen** [comment]
@reckart What about highlighting keywords in the annotation page, if no offsets are given?
